### PR TITLE
Run doctest for new files

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

The effect could be see in [this run](https://app.circleci.com/pipelines/github/huggingface/transformers/70724/workflows/f9ba5d54-6d64-4630-b500-ba1cf28baade/jobs/887892/artifacts) (see the artifact).

From Sir @ArthurZucker 

> new files should always be tested

> the documentation_tests.txt is only for legacy models taht are not tested

> but whenever someone adds something, (a new pipeline, a new model etc) should be automatic

### CODE REVIEW

File: utils/tests_fetcher.py [hunk 1]
```diff
@@ -404,8 +408,11 @@ def get_doctest_files(diff_with_last_commit: bool = False) -> List[str]:
         slow_documentation_tests = set(fp.read().strip().split("\n"))
 
     # So far we don't have 100% coverage for doctest. This line will be removed once we achieve 100%.
+    # TODO: handle renamed files correctly
     test_files_to_run = [
-        x for x in test_files_to_run if x in documentation_tests and x not in slow_documentation_tests
+        x
+        for x in test_files_to_run
+        if x in new_files or (x in documentation_tests and x not in slow_documentation_tests)
```
Comment (ydshieh) [inline utils/tests_fetcher.py:415]: new files are always added

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. It would be better to either:
a) reverse the list in the doctest file to include the files not tested yet (cause the docstrings haven't been fixed yet) in which case new files are automatically tested

or

b) put a check in the utils that checks the doctest file is properly sorted to make sure it includes new files

Either way, the new files would be always tested afterward (whereas here they are only tested when added). My preference goes for a if it's not too much work.

**Comment by ydshieh** [comment]
@ArthurZucker also mentioned `a` to me. I am happy with that, but we have to be a bit careful not to include files that has no doc example in that opposite list (if we want to use that list as an indication of **missing** doctest - it depends what we interpret it and use it)

File: utils/tests_fetcher.py [hunk 2]
```diff
@@ -396,16 +396,16 @@ def get_doctest_files(diff_with_last_commit: bool = False) -> List[str]:
...
+    with open("utils/not_doctested.txt") as fp:
+        not_doctested = set(fp.read().strip().split("\n"))
     # Do not run slow doctest tests
     with open("utils/slow_documentation_tests.txt") as fp:
         slow_documentation_tests = set(fp.read().strip().split("\n"))
 
     # So far we don't have 100% coverage for doctest. This line will be removed once we achieve 100%.
     test_files_to_run = [
-        x for x in test_files_to_run if x in documentation_tests and x not in slow_documentation_tests
+        x for x in test_files_to_run if x not in not_doctested and x not in slow_documentation_tests
```
Comment (ydshieh) [inline utils/tests_fetcher.py:408]: The files `slow_documentation_tests` won't be in `not_doctested` - they are doctested, but only on daily CI.

**Comment by ydshieh** [comment]
Added a `not_doctested.txt`. This list needs some more work, in particular `src/transformers/pipelines`. We should remove them out from this file so we can test them - but we have to make sure they work first!

(so far on `main`, they are not doctested)

I would prefer to separate that work from this PR.

File: utils/tests_fetcher.py (hunk 2 above)
Comment (ydshieh) [reply utils/tests_fetcher.py]: once PR approved, I will remove the file `documentation_list.txt`

**Review by ArthurZucker — COMMENTED** [review]
don't think that `init` files for example have anydoctest, so should probably not be here, and would prefer if only files that have doc example be here! (tests/* probably don't no?)

File: utils/not_doctested.txt [hunk 3]
```diff
@@ -0,0 +1,1593 @@
...
+docs/source/pt/pipeline_tutorial.md
+docs/source/pt/quicktour.md
+docs/source/pt/run_scripts.md
+docs/source/pt/serialization.md
+docs/source/pt/tasks/sequence_classification.md
+docs/source/pt/tasks/token_classification.md
+docs/source/pt/training.md
+docs/source/zh/index.md
+docs/source/zh/quicktour.md
+examples/README.md
```
Comment (sgugger) [inline utils/not_doctested.txt:499]: You should ignore the examples folder.

File: utils/not_doctested.txt [hunk 4]
```diff
@@ -0,0 +1,1593 @@
...
+src/transformers/utils/hub.py
+src/transformers/utils/import_utils.py
+src/transformers/utils/logging.py
+src/transformers/utils/model_parallel_utils.py
+src/transformers/utils/notebook.py
+src/transformers/utils/quantization_config.py
+src/transformers/utils/sentencepiece_model_pb2.py
+src/transformers/utils/sentencepiece_model_pb2_new.py
+src/transformers/utils/versions.py
+templates/adding_a_missing_tokenization_test/README.md
```
Comment (sgugger) [inline utils/not_doctested.txt:1534]: You should ignore the templates folder.

File: utils/not_doctested.txt [hunk 5]
```diff
@@ -0,0 +1,1593 @@
...
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/test_modeling_flax_{{cookiecutter.lowercase_modelname}}.py
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/test_modeling_tf_{{cookiecutter.lowercase_modelname}}.py
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/test_modeling_{{cookiecutter.lowercase_modelname}}.py
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/to_replace_{{cookiecutter.lowercase_modelname}}.py
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/tokenization_fast_{{cookiecutter.lowercase_modelname}}.py
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/tokenization_{{cookiecutter.lowercase_modelname}}.py
+templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/{{cookiecutter.lowercase_modelname}}.md
+templates/adding_a_new_model/open_model_proposals/ADD_BIG_BIRD.md
+templates/adding_a_new_model/open_model_proposals/README.md
+utils/add_pipeline_model_mapping_to_test.py
```
Comment (sgugger) [inline utils/not_doctested.txt:1553]: You should ignore the utils folder.

**Review by sgugger — COMMENTED** [review]
Thanks for diving into this! I think we can have stricter rules: only look in the english doc for md files and only in the src fodler for .py files. Everything else will never be doc-tested, and this will make the not_doctested file easier to read.

Also please include this file in the doctest_file check (so it's always sorted in alphabetical order).

File: utils/check_doctest_list.py [hunk 6]
```diff
@@ -37,7 +37,7 @@
 # All paths are set with the intent you should run this script from the root of the repo with the command
 # python utils/check_doctest_list.py
 REPO_PATH = "."
-DOCTEST_FILE_PATHS = ["documentation_tests.txt", "slow_documentation_tests.txt"]
```
Comment (ydshieh) [inline utils/check_doctest_list.py:40]: `documentation_tests.txt` will be removed before merge

File: utils/check_doctest_list.py [hunk 7]
```diff
@@ -62,7 +62,7 @@ def clean_doctest_list(doctest_file: str, overwrite: bool = False):
 
     if len(non_existent_paths) > 0:
         non_existent_paths = "\n".join([f"- {f}" for f in non_existent_paths])
-        raise ValueError(f"`utils/documentation_tests.txt` contains non-existent paths:\n{non_existent_paths}")
+        raise ValueError(f"`{doctest_file}` contains non-existent paths:\n{non_existent_paths}")
```
Comment (ydshieh) [inline utils/check_doctest_list.py:65]: just a miss

**Review by ArthurZucker — APPROVED** [review]
Much better! Fine for me 😉

**Review by sgugger — APPROVED** [review]
Thanks for iterating!

**Comment by ydshieh** [comment]
I will recreate the file today (so make sure it's updated with the latest `main`) and merge.

**Comment by ydshieh** [comment]
I am going to merge this PR without removing `utils/documentation_tests.txt` yet. It's still used by the doctest workflow file, and changing that file to use the opposite logic needs some extra work.

Will do it in a follow up PR.

**Pull request merged**